Refactor saving jit::Module to mobile .pt in 2 steps:

qihqi opened the pull request:
1. is to convert Function -> mobile::Function
2. is to serialize mobile::Function

This also opens opportunity to create mobile::Module without saving/reloading

Fixes #{issue number}

zhxchen17 commented on torch/csrc/jit/mobile/interpreter.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <torch/csrc/jit/mobile/interpreter.h>
+#include <iostream>
```
forgot to remove?

zhxchen17 commented on torch/csrc/jit/mobile/function.h:
```diff
@@ -56,6 +63,11 @@ class TORCH_API Function : public torch::jit::Function {
 
+std::function<void(Stack&)> makeOperatorFunction(
+    c10::OperatorName opname,
+    const c10::optional<int>& num_specified_args,
```
c10::optional<int>

zhxchen17 commented on torch/csrc/jit/serialization/export_bytecode.cpp:
```diff
@@ -1,11 +1,331 @@
+std::vector<Method> gatherMethodsToExport(
+    const Module& module,
+    bool incl_setstate,
+    bool incl_interface_call) {
```
I think we don't really want to parametrize this, we're gonna remove this soon, so keep it as global looks simpler.

qihqi replied:
true. Although to refer the global would need to include the function declared in export_modules, and introduce a circular dependency. We can fix up when we remove it.

zhxchen17 commented on torch/csrc/jit/serialization/export_bytecode.cpp:
```diff
@@ -1,11 +1,331 @@
+    }
+  }
+
+  // After getting the initial set of functions to export;
+  // recursively search for those that these function calls.
+  // BFS to cover ALL methods of submodules that are used
+  std::set<std::pair<std::string, Function*>> visited;
```
I like the idea that gatherMethodsToExport and compileGraphToMobileCode is separated. Here the problem is to find all interface calls, we need to inline the graph for once, and dropped the inlined graph, then when we actually compile it, we inline it again with the same result. This is probably fine, but I think in general we should only do the needed work exactly once. That's one of the reason why ExportedBytecodeSet is there, also to cache inlined graph.

I'd suggest we reuse and move out the private `exportFunction` from `export_module.cpp` which should already handle these things. More specially, my "ideal" gatherMethodsToExport is like:
```
def gatherMethodsToExport() {
  BytecodeExportSet exportSet;
  for (method : module.methods) {
    exportFunction(exportSet, method);
  }
}
```
Now if you don't like recursive `exportFunction`, another option is still keeping BFS, but at least leverage cached graphs, like:
```
def gatherMethodsToExport() {
  BytecodeExportSet exportSet;
  // BFS loop
  while (...) {
    // ...
    exportSet.add(stack_top, cached_graph);
    // ...
  }
}
```

zhxchen17 commented on torch/csrc/jit/serialization/export_bytecode.cpp:
```diff
@@ -1,11 +1,331 @@
+
+std::unique_ptr<mobile::Code> compileGraphToMobileCode(
+    const std::string& name,
+    const std::shared_ptr<Graph>& graph,
```
It'd great if we also update `MobileCode` to only take `const Graph&`, but that can be in a separate diff.

zhxchen17 commented on torch/csrc/jit/serialization/export_bytecode.h:
```diff
@@ -16,9 +20,11 @@ struct ExportedFunction {
       const Module& m,
       const Function& f,
-      std::unique_ptr<Graph> g,
+      std::unique_ptr<Graph>&& g,
```
We don't conditionally move the graph here, because we always move it. Usually `&&` express the intension that an argument maybe moved or not.

zhxchen17 commented on torch/csrc/jit/serialization/export_bytecode.cpp:
```diff
@@ -1,11 +1,331 @@
+
+  std::unique_ptr<mobile::Code> mobile_code_ptr =
+      std::make_unique<mobile::Code>();
+  mobile::Code& mobile_code = *mobile_code_ptr.get();
```
nit: *mobile_code;

zhxchen17 reviewed (commented):
Overall this diff makes sense to me. Should we rename the title to be something like "Refactor saving Functions from jit::Module to mobile"?

zhxchen17 commented on torch/csrc/jit/serialization/export_bytecode.cpp:
```diff
@@ -1,23 +1,322 @@
+    c10::QualifiedName qn(*cur.owner()._ivalue()->type()->name(), f.name());
+
+    if (incl_dependent_functions) {
+      std::unordered_set<c10::string_view> called_method_names;
+      auto nodes = findAllNodes(*graph, c10::prim::CallMethod, true);
+      for (Node* node : nodes) {
+        if (auto iface = node->input(0)->type()->castRaw<InterfaceType>()) {
+          const FunctionSchema* schema = iface->getMethod(node->s(attr::name));
+          called_method_names.insert(schema->name());
+        }
+      }
+
+      for (const auto& submodule : cur.owner().modules()) {
+        for (const auto& m : submodule.get_methods()) {
+          if (called_method_names.find(m.function().qualname().name()) !=
+              called_method_names.end()) {
```
@qihqi Could we put this in a separate function? (like vector<> findInterfaceMethods()).

qihqi replied:
done.

iseeyuan reviewed (commented):
There are both refactoring and updates in this PR. It would be great to split this PR to two, one for refactoring only, and one for updates only. It would help a lot for both record tracking and code review. Thanks!

qihqi commented on torch/csrc/jit/serialization/export_bytecode.cpp:
```diff
@@ -1,23 +1,333 @@
+      mobile_code.op_names_.emplace_back(opname);
+      auto func = mobile::makeOperatorFunction(
+          opname, num_args, compilation_options.model_version);
+      TORCH_INTERNAL_ASSERT(
```
We are not exporting yet. 

This is converting Graph -> bytecode in memory. So the operator has to exist. The operator exists because it is bound in the Graph's Node already. Alternatively we could use the Operator* inside of Node instead of doing an lookup with name; but then we still need something to wrap it to std::function so the extra saving of one less lookup is not worth it.

qihqi commented:
> There are both refactoring and updates in this PR. It would be great to split this PR to two, one for refactoring only, and one for updates only. It would help a lot for both record tracking and code review. Thanks!

Sorry what is the "update" part? My intent is refactor only.

iseeyuan commented:
> Sorry what is the "update" part? My intent is refactor only.

For "refactor" I meant moving codes from a file to another, or out of a big function, without any changes inside the code content (if it includes moving to a new file, necessary build inclusion is added). "Update" refers code changes in addition to the code movement. For example, function `makeOperatorFunction` is added to the Function class.

qihqi commented:
> > Sorry what is the "update" part? My intent is refactor only.
> 
> For "refactor" I meant moving codes from a file to another, or out of a big function, without any changes inside the code content (if it includes moving to a new file, necessary build inclusion is added). "Update" refers code changes in addition to the code movement. For example, function `makeOperatorFunction` is added to the Function class.

I see make sense! `makeOperatorFunction` is probably a bad example because it is copy & pasted out from append_operator(), then wrapped into a new function that append_operator() calls; but yeah your point stands. there are +200 net non-test LOCs in this diff and probably can be split into separate commits.

qihqi committed:
Refactor saving jit::Module to mobile .pt in 2 steps: (#66494)

Summary:
1. is to convert Function -> mobile::Function
2. is to serialize mobile::Function

This also opens opportunity to create mobile::Module without saving/reloading

Fixes #{issue number}

Pull Request resolved: https://github.com/pytorch/pytorch/pull/66494

Reviewed By: zhxchen17

Differential Revision: D32293022

Pulled By: qihqi

fbshipit-source-id: 5e9c0fc9ba9f010f963500cc9328cf8211215b45
